### ISSUE TITLE
Model reproduction issue - lower performance than expected

### ISSUE THREAD

**Issue by sam-james-harding** [hanjun815/SHeRLoc#3]
Hi,

I have been attempting to reproduce the results from the SHeRLoc paper and have been having some issues with getting lower recall values than expected. 

Here is the log from one attempt at training:

```
Training config path: ../config/config.txt
Model config path: ../models/Hetero_SHeRLoc.txt
Debug mode: False
Parameters:
params_path: ../config/config.txt
model_params_path: ../models/Hetero_SHeRLoc.txt
dataset: hercules
dataset_folder: <HERCULES_DIRECTORY>/HeRCULES
num_workers: 8
batch_size: 256
lr: 0.001
epochs: 20
weight_decay: 0.001
loss: BatchHardTripletMarginLoss
margin: 1.0
train_dataset: <HERCULES_DIRECTORY>/HeRCULES
val_dataset: <HERCULES_DIRECTORY>/HeRCULES
test_dataset: <HERCULES_DIRECTORY>/HeRCULES
Model parameters:
model_params_path: ../models/Hetero_SHeRLoc.txt
task: Hetero
radar_model: SHeRLoc


Model name: model_SHeRLoc_20260120_1216
Number of model parameters: 19529301
11710 readings from sensor: png with valid pose, 0 rejected due to unknown pose
2342 readings from sensor: png with valid pose, 0 rejected due to unknown pose
9319 readings from sensor: png with valid pose, 0 rejected due to unknown pose
1864 readings from sensor: png with valid pose, 0 rejected due to unknown pose
7805 readings from sensor: png with valid pose, 0 rejected due to unknown pose
1561 readings from sensor: png with valid pose, 0 rejected due to unknown pose
11630 readings from sensor: png with valid pose, 0 rejected due to unknown pose
2331 readings from sensor: png with valid pose, 0 rejected due to unknown pose
19575 readings from sensor: png with valid pose, 0 rejected due to unknown pose
3930 readings from sensor: png with valid pose, 0 rejected due to unknown pose
2510 readings from sensor: png with valid pose, 0 rejected due to unknown pose
502 readings from sensor: png with valid pose, 0 rejected due to unknown pose
2237 readings from sensor: png with valid pose, 0 rejected due to unknown pose
447 readings from sensor: png with valid pose, 0 rejected due to unknown pose
2321 readings from sensor: png with valid pose, 0 rejected due to unknown pose
464 readings from sensor: png with valid pose, 0 rejected due to unknown pose
2198 readings from sensor: png with valid pose, 0 rejected due to unknown pose
440 readings from sensor: png with valid pose, 0 rejected due to unknown pose

  0%|          | 0/20 [00:00<?, ?it/s]0 45360 0.25886547565460205

train - Global loss: 0.202814    Embedding norm: 1.0000   Triplets (all/active): 254.8/28159.5
Pos dist (min/mean/max): 0.4535/0.8693/1.3976   Neg dist (min/mean/max): 0.6750/1.3995/1.8460

  5%|▌         | 1/20 [04:20<1:22:31, 260.63s/it]0 2020 0.2661606967449188

val - Global loss: 0.224604    Embedding norm: 1.0000   Triplets (all/active): 252.5/36834.4
Pos dist (min/mean/max): 0.5696/0.9872/1.4141   Neg dist (min/mean/max): 0.6983/1.4106/1.8038

train - Global loss: 0.129965    Embedding norm: 1.0000   Triplets (all/active): 254.8/17712.3
Pos dist (min/mean/max): 0.5121/0.8575/1.2789   Neg dist (min/mean/max): 0.8293/1.4087/1.7860

 10%|█         | 2/20 [08:38<1:17:44, 259.15s/it]0 2020 0.2714903652667999

val - Global loss: 0.206885    Embedding norm: 1.0000   Triplets (all/active): 252.5/37736.6
Pos dist (min/mean/max): 0.5884/1.0070/1.3960   Neg dist (min/mean/max): 0.6204/1.4151/1.7822

train - Global loss: 0.103231    Embedding norm: 1.0000   Triplets (all/active): 254.8/12012.1
Pos dist (min/mean/max): 0.5034/0.8141/1.1903   Neg dist (min/mean/max): 0.8625/1.4092/1.7722

 15%|█▌        | 3/20 [12:57<1:13:19, 258.81s/it]0 2020 0.1689797043800354

val - Global loss: 0.196322    Embedding norm: 1.0000   Triplets (all/active): 252.5/24875.6
Pos dist (min/mean/max): 0.4386/0.8876/1.4218   Neg dist (min/mean/max): 0.4536/1.3706/1.7855

train - Global loss: 0.088278    Embedding norm: 1.0000   Triplets (all/active): 254.8/9675.1
Pos dist (min/mean/max): 0.5002/0.7884/1.1362   Neg dist (min/mean/max): 0.8828/1.4094/1.7691

 20%|██        | 4/20 [17:16<1:09:01, 258.86s/it]0 2020 0.19869627058506012

val - Global loss: 0.174237    Embedding norm: 1.0000   Triplets (all/active): 252.5/25094.8
Pos dist (min/mean/max): 0.5634/0.9872/1.3428   Neg dist (min/mean/max): 0.6595/1.4375/1.7573

train - Global loss: 0.075101    Embedding norm: 1.0000   Triplets (all/active): 254.8/6611.6
Pos dist (min/mean/max): 0.4911/0.7525/1.0827   Neg dist (min/mean/max): 0.8947/1.4091/1.7599

 25%|██▌       | 5/20 [21:34<1:04:40, 258.68s/it]0 2020 0.12267506867647171

val - Global loss: 0.136666    Embedding norm: 1.0000   Triplets (all/active): 252.5/15793.0
Pos dist (min/mean/max): 0.6169/0.9266/1.3407   Neg dist (min/mean/max): 0.7307/1.4333/1.7751

train - Global loss: 0.068319    Embedding norm: 1.0000   Triplets (all/active): 254.8/5614.1
Pos dist (min/mean/max): 0.4855/0.7413/1.0583   Neg dist (min/mean/max): 0.9224/1.4092/1.7540

 30%|███       | 6/20 [25:53<1:00:23, 258.84s/it]0 2020 0.13457196950912476

val - Global loss: 0.162283    Embedding norm: 1.0000   Triplets (all/active): 252.5/15613.2
Pos dist (min/mean/max): 0.5364/0.9159/1.3346   Neg dist (min/mean/max): 0.7145/1.4448/1.7969

train - Global loss: 0.061000    Embedding norm: 1.0000   Triplets (all/active): 254.8/4729.8
Pos dist (min/mean/max): 0.4808/0.7251/1.0205   Neg dist (min/mean/max): 0.9203/1.4090/1.7501

 35%|███▌      | 7/20 [30:12<56:05, 258.85s/it]  0 2020 0.15395431220531464

val - Global loss: 0.156209    Embedding norm: 1.0000   Triplets (all/active): 252.5/16828.4
Pos dist (min/mean/max): 0.5536/0.9517/1.3564   Neg dist (min/mean/max): 0.6796/1.4372/1.7564

train - Global loss: 0.054510    Embedding norm: 1.0000   Triplets (all/active): 254.8/4355.5
Pos dist (min/mean/max): 0.4814/0.7171/1.0032   Neg dist (min/mean/max): 0.9363/1.4090/1.7453

 40%|████      | 8/20 [34:31<51:48, 259.01s/it]0 2020 0.18811602890491486

val - Global loss: 0.158739    Embedding norm: 1.0000   Triplets (all/active): 252.5/22590.4
Pos dist (min/mean/max): 0.5898/0.9912/1.3662   Neg dist (min/mean/max): 0.6959/1.4522/1.7734

train - Global loss: 0.048540    Embedding norm: 1.0000   Triplets (all/active): 254.8/3641.2
Pos dist (min/mean/max): 0.4799/0.7056/0.9773   Neg dist (min/mean/max): 0.9519/1.4094/1.7402

 45%|████▌     | 9/20 [38:50<47:27, 258.90s/it]0 2020 0.17865116894245148

val - Global loss: 0.163757    Embedding norm: 1.0000   Triplets (all/active): 252.5/16869.1
Pos dist (min/mean/max): 0.5815/0.9328/1.3284   Neg dist (min/mean/max): 0.6398/1.4367/1.7626

train - Global loss: 0.042091    Embedding norm: 1.0000   Triplets (all/active): 254.8/2909.7
Pos dist (min/mean/max): 0.4696/0.6888/0.9552   Neg dist (min/mean/max): 0.9547/1.4090/1.7365

 50%|█████     | 10/20 [43:10<43:11, 259.14s/it]0 2020 0.14572566747665405

val - Global loss: 0.141928    Embedding norm: 1.0000   Triplets (all/active): 252.5/14588.2
Pos dist (min/mean/max): 0.6375/0.9551/1.3475   Neg dist (min/mean/max): 0.7302/1.4423/1.7340

train - Global loss: 0.038536    Embedding norm: 1.0000   Triplets (all/active): 254.8/2624.9
Pos dist (min/mean/max): 0.4600/0.6804/0.9307   Neg dist (min/mean/max): 0.9572/1.4091/1.7360

 55%|█████▌    | 11/20 [47:29<38:52, 259.18s/it]0 2020 0.1745242327451706

val - Global loss: 0.147424    Embedding norm: 1.0000   Triplets (all/active): 252.5/16013.8
Pos dist (min/mean/max): 0.5487/0.9077/1.2931   Neg dist (min/mean/max): 0.6830/1.4146/1.7408

train - Global loss: 0.032807    Embedding norm: 1.0000   Triplets (all/active): 254.8/1963.1
Pos dist (min/mean/max): 0.4618/0.6706/0.9152   Neg dist (min/mean/max): 0.9690/1.4084/1.7322

 60%|██████    | 12/20 [51:48<34:34, 259.28s/it]0 2020 0.15196670591831207

val - Global loss: 0.134580    Embedding norm: 1.0000   Triplets (all/active): 252.5/14705.1
Pos dist (min/mean/max): 0.6358/0.9525/1.3097   Neg dist (min/mean/max): 0.7533/1.4422/1.7884

train - Global loss: 0.027236    Embedding norm: 1.0000   Triplets (all/active): 254.8/1677.7
Pos dist (min/mean/max): 0.4618/0.6662/0.9015   Neg dist (min/mean/max): 0.9756/1.4088/1.7257

 65%|██████▌   | 13/20 [56:08<30:14, 259.26s/it]0 2020 0.18635132908821106

val - Global loss: 0.133940    Embedding norm: 1.0000   Triplets (all/active): 252.5/16869.6
Pos dist (min/mean/max): 0.6293/0.9921/1.3082   Neg dist (min/mean/max): 0.7176/1.4585/1.7446

train - Global loss: 0.025545    Embedding norm: 1.0000   Triplets (all/active): 254.8/1470.7
Pos dist (min/mean/max): 0.4653/0.6617/0.8905   Neg dist (min/mean/max): 0.9956/1.4092/1.7280

 70%|███████   | 14/20 [1:00:26<25:54, 259.07s/it]0 2020 0.17590665817260742

val - Global loss: 0.136447    Embedding norm: 1.0000   Triplets (all/active): 252.5/18646.8
Pos dist (min/mean/max): 0.5811/0.9681/1.3074   Neg dist (min/mean/max): 0.7671/1.4336/1.7453

train - Global loss: 0.018942    Embedding norm: 1.0000   Triplets (all/active): 254.8/1218.1
Pos dist (min/mean/max): 0.4584/0.6501/0.8679   Neg dist (min/mean/max): 0.9947/1.4090/1.7281

 75%|███████▌  | 15/20 [1:04:45<21:35, 259.01s/it]0 2020 0.18802636861801147

val - Global loss: 0.128234    Embedding norm: 1.0000   Triplets (all/active): 252.5/16147.8
Pos dist (min/mean/max): 0.6110/0.9864/1.3049   Neg dist (min/mean/max): 0.7303/1.4554/1.7563

train - Global loss: 0.016234    Embedding norm: 1.0000   Triplets (all/active): 254.8/977.1
Pos dist (min/mean/max): 0.4556/0.6443/0.8534   Neg dist (min/mean/max): 0.9970/1.4093/1.7243

 80%|████████  | 16/20 [1:09:04<17:15, 258.86s/it]0 2020 0.13367825746536255

val - Global loss: 0.127193    Embedding norm: 1.0000   Triplets (all/active): 252.5/16167.9
Pos dist (min/mean/max): 0.5980/0.9638/1.3017   Neg dist (min/mean/max): 0.7103/1.4445/1.7547

train - Global loss: 0.013518    Embedding norm: 1.0000   Triplets (all/active): 254.8/768.7
Pos dist (min/mean/max): 0.4505/0.6360/0.8383   Neg dist (min/mean/max): 1.0023/1.4091/1.7220

 85%|████████▌ | 17/20 [1:13:22<12:55, 258.65s/it]0 2020 0.144673690199852

val - Global loss: 0.127432    Embedding norm: 1.0000   Triplets (all/active): 252.5/16017.9
Pos dist (min/mean/max): 0.6177/0.9773/1.2991   Neg dist (min/mean/max): 0.7843/1.4490/1.7682

train - Global loss: 0.011328    Embedding norm: 1.0000   Triplets (all/active): 254.8/608.3
Pos dist (min/mean/max): 0.4488/0.6327/0.8310   Neg dist (min/mean/max): 0.9998/1.4092/1.7243

 90%|█████████ | 18/20 [1:17:40<08:37, 258.65s/it]0 2020 0.15564939379692078

train - Global loss: 0.009287    Embedding norm: 1.0000   Triplets (all/active): 254.8/553.3
Pos dist (min/mean/max): 0.4460/0.6293/0.8234   Neg dist (min/mean/max): 1.0032/1.4091/1.7254

 95%|█████████▌| 19/20 [1:21:59<04:18, 258.56s/it]0 2020 0.16164810955524445

val - Global loss: 0.133025    Embedding norm: 1.0000   Triplets (all/active): 252.5/16867.8
Pos dist (min/mean/max): 0.6062/0.9838/1.3006   Neg dist (min/mean/max): 0.6991/1.4509/1.7685

train - Global loss: 0.008173    Embedding norm: 1.0000   Triplets (all/active): 254.8/507.7
Pos dist (min/mean/max): 0.4453/0.6277/0.8199   Neg dist (min/mean/max): 1.0063/1.4095/1.7260

100%|██████████| 20/20 [1:26:18<00:00, 258.63s/it]
100%|██████████| 20/20 [1:26:18<00:00, 258.91s/it]

val - Global loss: 0.121539    Embedding norm: 1.0000   Triplets (all/active): 252.5/16351.9
Pos dist (min/mean/max): 0.6063/0.9786/1.3008   Neg dist (min/mean/max): 0.7608/1.4527/1.7663
Training completed.
```

The results from this attempt at training (from the evaluation script, tested on Sports_Complex/01 for both the map and query sequence) are:
```
recall@ 1 :  57.03410475030451
recall@ 2 :  57.429963459196095
recall@ 3 :  58.92204628501827
recall@ 4 :  61.47990255785627
recall@ 5 :  62.9415347137637
recall@ 6 :  64.43361753958587
recall@ 7 :  65.89524969549329
recall@ 8 :  66.71741778319124
recall@ 9 :  67.87454323995128
recall@ 10 :  68.42265529841657
```

Another attempt under similar conditions had a global validation loss of 0.102397, and the following evaluation results:
```
recall@ 1 :  66.80876979293544
recall@ 2 :  66.90012180267966
recall@ 3 :  68.30085261875762
recall@ 4 :  70.58465286236297
recall@ 5 :  72.01583434835567
recall@ 6 :  73.8733252131547
recall@ 7 :  75.30450669914738
recall@ 8 :  75.94397076735689
recall@ 9 :  76.82704019488429
recall@ 10 :  77.557856272838
```

These are considerably better but still far from the expected results, and it's not clear why performance improved here anyway (results similarly varied in additional training attempts). The config.txt used for training is the following:

```
[DEFAULT]
dataset=HeRCULES
dataset_folder = <HERCULES DIRECTORY>/HeRCULES

[TRAIN]
num_workers = 8
batch_size = 256

lr = 1e-3
epochs = 20
weight_decay = 1e-3
loss = BatchHardTripletMarginLoss
margin = 1.0 

train_dataset = <HERCULES DIRECTORY>/HeRCULES
val_dataset = <HERCULES DIRECTORY>/HeRCULES
test_dataset = <HERCULES DIRECTORY>/HeRCULES
```

Increasing the number of epochs from 20 to 25 has also had negligible effect:
```
recall@ 1 :  55.66382460414129
recall@ 2 :  55.9074299634592
recall@ 3 :  58.526187576126674
recall@ 4 :  60.444579780755184
recall@ 5 :  62.45432399512789
recall@ 6 :  63.61144945188795
recall@ 7 :  64.6163215590743
recall@ 8 :  65.16443361753959
recall@ 9 :  65.92570036540803
recall@ 10 :  66.47381242387333
```

I have also looked into #2 and made sure that the batch size is 256. Training using 4 GPUs rather than a single GPU has also yielded very similar evaluation results to this (the GPUs being used are NVIDIA H100s).

I was wondering whether you might have any insight into how I can achieve similar results to those in the paper. Evaluation using the provided weights has produced very similar results to the paper, so it seems that the issue must be happening in the training stage. Also, the training data that was used is the pre-processed HeRCULES dataset provided at the link in the README of this repo.

**Comment by hanjun815** [comment]
Hi,

I haven’t encountered the same issue, but I recommend first verifying that the evaluation is performed exactly as intended. Please make sure you are running the evaluation with the following command:
```
python evaluate.py \
  --dataset_root /code/SHeRLoc/datasets/HeRCULES \
  --map_seq Sports_Complex/01 \
  --query_seq Sports_Complex/01 \
  --model_config ../models/Hetero_SHeRLoc.txt \
  --weights ../weights/Hetero_SHeRLoc.pth
```

If you have confirmed that the evaluation setup is correct and the issue still persists, it is likely related to the training process. In that case, the gap is certainly from numerical and kernel-level differences (TF32/FMA, NCCL reduction ordering), which significantly affect BatchHard triplet mining.

Fixes to try
(A) Make H100 run match FP32 determinism
(example)
```
export NVIDIA_TF32_OVERRIDE=0
```
(B) Sanity run with AMP off (FP32)
(C) Tune training hyperparameters for your setup (e.g., margin, learning rate)

Also, please check PyTorch/CUDA versions, whether AMP is on (bf16/fp16), and whether you use DDP.
